src/svm/eventsClient.ts

```diff
@@ -0,0 +1,205 @@
+    });
+    return new SvmSpokeEventsClient(rpc, svmSpokeAddress, svmSpokeEventAuthority);
+  }
+
+  /**
+   * Queries events for the SvmSpoke program filtered by event name.
+   *
+   * @param eventName - The name of the event to filter by.
+   * @param fromSlot - Optional starting slot.
+   * @param toSlot - Optional ending slot.
+   * @param options - Options for fetching signatures.
+   * @returns A promise that resolves to an array of events matching the eventName.
+   */
+  public async queryEvents<T extends EventData>(
+    eventName: EventName,
+    fromSlot?: bigint,
+    toSlot?: bigint,
+    options: GetSignaturesForAddressConfig = { limit: 1000, commitment: "confirmed" }
+  ): Promise<EventWithData<T>[]> {
+    const events = await this.queryAllEvents(fromSlot, toSlot, options);
+    return events.filter((event) => event.name === eventName) as EventWithData<T>[];
+  }
+
+  /**
+   * Queries all events for a specific program.
+   *
+   * @param fromSlot - Optional starting slot.
+   * @param toSlot - Optional ending slot.
+   * @param options - Options for fetching signatures.
+   * @returns A promise that resolves to an array of all events with additional metadata.
+   */
+  private async queryAllEvents(
+    fromSlot?: bigint,
+    toSlot?: bigint,
+    options: GetSignaturesForAddressConfig = { limit: 1000, commitment: "confirmed" }
+  ): Promise<EventWithData<EventData>[]> {
+    const allSignatures: GetSignaturesForAddressTransaction[] = [];
+    let hasMoreSignatures = true;
+    let currentOptions = options;
+
+    while (hasMoreSignatures) {
+      const signatures: GetSignaturesForAddressApiResponse = await this.rpc
+        .getSignaturesForAddress(this.svmSpokeAddress, currentOptions)
+        .send();
+      // Signatures are sorted by slot in descending order.
+      allSignatures.push(...signatures);
+
+      // Update options for the next batch. Set "before" to the last fetched signature.
+      if (signatures.length > 0) {
+        currentOptions = { ...currentOptions, before: signatures[signatures.length - 1].signature };
+      }
+
+      if (fromSlot && allSignatures.length > 0 && allSignatures[allSignatures.length - 1].slot < fromSlot) {
+        hasMoreSignatures = false;
+      }
+
+      hasMoreSignatures = Boolean(
+        hasMoreSignatures && currentOptions.limit && signatures.length === currentOptions.limit
+      );
+    }
+
+    const filteredSignatures = allSignatures.filter((signatureTransaction) => {
+      if (fromSlot && signatureTransaction.slot < fromSlot) return false;
+      if (toSlot && signatureTransaction.slot > toSlot) return false;
+      return true;
+    });
+
+    // Fetch events for all signatures in parallel.
+    const eventsWithSlots = await Promise.all(
+      filteredSignatures.map(async (signatureTransaction) => {
+        const events = await this.readEventsFromSignature(signatureTransaction.signature, options.commitment);
+        return events.map((event) => ({
```

Comment:
you can add more info to these events such as the name of the event, which is useful, from `event.name` ([see](https://github.com/across-protocol/contracts/blob/4d89becb02ddd6277b1055dbe7561bd35f334dee/src/svm/web3-v1/solanaProgramUtils.ts#L168)). 

also do you think it's better to have `confirmationStatus` &`blockTime` defined as a string/UnixTimestamp or undefined or better to set them to hard limits so the types are always string & UnixTimestamp as done [here](https://github.com/across-protocol/contracts/blob/4d89becb02ddd6277b1055dbe7561bd35f334dee/src/svm/web3-v1/solanaProgramUtils.ts#L164)?

Reply:
Yep, the event name and parsed data, which match the Codama clients’ generated event types, all come from [this function](https://github.com/across-protocol/sdk/blob/f3ed1129ca2d3a56004c39d0a602c8a906a39e0b/src/svm/eventsClient.ts#L159-L201) and included when we destructure `...event`

Reply:
Regarding the hard limits, I agreed with Reinis’ suggestion to use `null` as they are already defined in solana web3 ([link](https://github.com/across-protocol/sdk/pull/899#discussion_r1975259824)).
It feels more natural to keep the original outputs and let the consumer decide, but I’m not strongly opinionated on this.